Refactor STM32 PWM driver using LL API

gmarull opened the pull request:
The PWM driver has been refactored using the HAL LL API. Not only that, but the set pin_set function is now faster, as channel output compare is just initialized if needed.

Support for the polarity flag has also been added to the STM32 PWM driver. STM32 boards using PWM have been updated accordingly.

Fixes #23629

gmarull committed:
drivers: pwm: stm32: refactor driver using LL API

The PWM drivers has been refactored using the HAL LL API. Not only that,
but the set pin_set function is now faster, as channel output compare is
just initialized if needed.

NOTE: Has been tested using H743zi board for now.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented on boards/arm/nucleo_h743zi/pinmux.c:
```diff
@@ -22,9 +22,9 @@ static const struct pin_config pinconf[] = {
 	{ STM32_PIN_PB9, STM32H7_PINMUX_FUNC_PB9_I2C1_SDA },
 #endif
-#ifdef CONFIG_PWM_STM32_12
+#if DT_HAS_NODE(DT_NODELABEL(timers12))
```
This needs a better solution (timer is now only used for PWM, but that may change). I'd like to hear proposals, maybe labeling each PWM with `pwm{n}`?

henrikbrixandersen replied:
I have submitted a PR for solving a similar issue for the NXP FlexTimer (which is currently only used for PWM in Zephyr, but has many uses - some, which I am working on providing drivers for) in #24498. My proposed solution is similar to that used for handling the different use-cases for the NXP Flexcomm instances in Zephyr.

gmarull replied:
Thanks @henrikbrixandersen. It looks like a good approach that can be applied here. If I get it right the functionality is selected by specifying which driver (`compatible`) will use that peripheral.

gmarull replied:
FYI: This is how it is handled on U-Boot (same in Kernel):
https://elixir.bootlin.com/u-boot/latest/source/arch/arm/dts/stm32mp157c.dtsi#L132
https://elixir.bootlin.com/u-boot/latest/source/arch/arm/dts/stm32mp157c-ev1.dts#L300.

henrikbrixandersen replied:
> If I get it right the functionality is selected by specifying which driver (`compatible`) will use that peripheral.

@gmarull That is correct.

gmarull replied:
@henrikbrixandersen This article may be useful as it describes in detail how Linux handles the same problem https://bootlin.com/blog/timer-counters-linux-microchip/ (what a coincidence, it was posted today :smile: )

erwango replied:
@gmarull on my side, I'd prefer to go with the label addition as I think was your initial proposal (same as what is describe in the article you mention finaly). This  would help to stay close to stm32 linux.

henrikbrixandersen commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -16,26 +15,108 @@
+{
+	if (flags & PWM_POLARITY_NORMAL) {
+		return LL_TIM_OCPOLARITY_HIGH;
+	}
```
When you add support for PWM flags here, you should also add support for flags to the PWM DTS binding (see https://github.com/zephyrproject-rtos/zephyr/blob/ccb5a01fc18ca5e8879cfa21adbb77b9740a6ac5/dts/bindings/pwm/nxp%2Ckinetis-ftm.yaml#L24 for example).

gmarull replied:
Thanks @henrikbrixandersen! I have pushed a new commit with updated DT bindings as well as affected boards.

gmarull committed:
drivers: pwm: stm32: add support for polarity

Add support for the polarity flag in the STM32 PWM driver.

STM32 boards using PWM have been updated accordingly.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

erwango commented on drivers/pwm/Kconfig.stm32:
```diff
@@ -3,135 +3,11 @@
 	  processors. Say y if you wish to use PWM port on STM32
 	  MCU.
-
-if PWM_STM32
-
-config PWM_STM32_1
-	bool "STM32 PWM 1 Output"
-	help
-	  Enable output for PWM1 in the driver. Say y here
-	  if you want to use PWM1 output.
-
```
Removing these symbols require to update whole set of boards using them.
Please do the board change accordingly.

gmarull replied:
Thanks, will do. I preferred to wait until an initial review before doing that as noted in the PR notes. Before doing that, I'd also like your feedback on the question above (regarding `DT_HAS_NODE(DT_NODELABEL(timers12))`).

erwango replied:
Oups I didn't want this to be put as a single comment.
So indeed, we need to sort the labeling issue first.

erwango reviewed (changes requested):
This PR is ambitious has it encompass 2 major changes: Boards Kconfig to dts transition and conversion to LL, plus and additional rather unrelated change (polarity support).
Please split these 3 changes to individual PRs.

Regarding the Kconfig to dts transition, I think the node labeling sound like the best option:
```
		timers1: timers@40012c00 {
			compatible = "st,stm32-timers";
			{...}                  
			pmw1: pwm@1 {
				compatible = "st,stm32-pwm";
				status = "disabled";
				st,prescaler = <10000>;
				label = "PWM_1";
				#pwm-cells = <2>;
			};
```
Besides, it would help getting rid of the paths in following expressions:
```
		green_pwm_led: green_pwm_led {
			pwms = <&{/soc/timers@40000000/pwm} 1 4>;
```

gmarull commented:
@erwango I have rebased this PR now that PWM uses DT. I have splitted it in two commits, the first one is the refactor and the second the addition of the polarity flag support. I can split it into 2 PRs, not sure if needed, though.

FRASTM commented on dts/bindings/pwm/st,stm32-pwm.yaml:
```diff
@@ -14,9 +14,10 @@ properties:
   - channel
 # period in terms of nanoseconds
   - period
+  - flags
```
flags is either `PWM_POLARITY_NORMAL` or  `PWM_POLARITY_INVERTED` , correct ?

gmarull replied:
Yes, they're defined `#include <dt-bindings/pwm/pwm.h>`. I guess that flags could potentially be used for other purposes in the future.

gmarull commented:
@erwango ping

erwango reviewed (changes requested):
Txs for this welcome change.
Some comments, but otherwise it's fine to me.

gmarull commented on boards/arm/nucleo_f401re/nucleo_f401re.dts:
```diff
@@ -8,6 +8,7 @@
 /dts-v1/;
 #include <st/f4/stm32f401Xe.dtsi>
 #include "arduino_r3_connector.dtsi"
+#include <dt-bindings/pwm/pwm.h>
```
Done, thanks

gmarull commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -16,26 +15,96 @@
 
 #include <drivers/clock_control/stm32_clock_control.h>
 
-#include "pwm_stm32.h"
```
Good catch, deleted.

gmarull commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -1,11 +1,10 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#define DT_DRV_COMPAT st_stm32_pwm
```
Added to the top. The reason for doing that is to prevent accidental direct usage of DT macros, as it usually makes drivers not capable of handling an arbitrary number of instances.

erwango replied:
As mentioned, I'm totally open to change this but it would be better to discuss and update it widely then.

lochej commented:
I'm sorry if this is the wrong section but that is my first time contributing to Zephyr. 
The LL PWM implementation has a bug in the **get_polarity()** function always returns **LL_TIM_OCPOLARITY_LOW** which in turn actually inverts your PWM on STM32s :

```
static uint32_t get_polarity(pwm_flags_t flags)
{
	if (flags & PWM_POLARITY_NORMAL) { <-- PWM_POLARITY_NORMAL expands to 0<<0 in include/dt-bindings/pwm/pwm.h
		return LL_TIM_OCPOLARITY_HIGH;
	}

	return LL_TIM_OCPOLARITY_LOW;
}
```

The **PWM_POLARITY_MASK** set to 0<<0 forces the **if** statement to always be false.

My recommendation would either be to change the defines in **dt-bindings/pwm/pwm.h** to:
```
/** PWM pin normal polarity (active-high pulse). */
#define PWM_POLARITY_NORMAL	(1 << 0)
// -> High polarity should be 1

/** PWM pin inverted polarity (active-low pulse). */
#define PWM_POLARITY_INVERTED	(0 << 0)
//-> Low polarity should be 0
```
Doing this fix has proven correct behavior of PWM polarity on my Nucleo64 board.

gmarull commented:
@lochej thanks for reporting, fix here: https://github.com/zephyrproject-rtos/zephyr/pull/26680. You can open an issue if you want so it can be referenced on the fix PR.
